### ISSUE TITLE
Make it easier to find all EKS dependencies

### ISSUE THREAD

**Issue by lancehunt** [aws/eks-distro#72]
**What would you like to be added**:
Documentation to understand all adjacent dependencies of an EKS cluster, such as (not comprehensive list):

1. EKS AMI Building (https://github.com/awslabs/amazon-eks-ami) 
2. Amazon Linux 2 used in AMI building (for hardening)
3. AWS CNI
4. Etcd Version
5. CoreDns version

Honestly, I dont know how far down the rabbit hole goes, but thats the point.

**Why is this needed**:

As a EKS cluster creator, cloud-architect, security professional or curious enthusiast, I need to understand the choices made in EKS that could be material to my own decision-making.

Examples are:
- Ability to experiment with new patches to AWS CNI or other components before officially adopted
- Determine if a Linux, Etcd, or other component's CVE applies to my EKS cluster
- Determine whether EKS AMI's meet CIS standards
- Determine whether Linux Kernel features are available to Nodes and Containers within the cluster (kernel 5.4 is not supported by Amazon Linux 2 which is a dependency of EKS AMI's which is a dependency of EKS)

Some of these are very specific to needs I have encountered, but I can see benefits of good documentation of all dependencies and links to adjacent Github repos where further details could be found on secondary and tertiary dependencies.

Basically, don't make me go hunting through this repo and several others to understand the totality of all EKS dependencies.

**Comment by mhausenblas** [comment]
Thanks for raising this issue and I'd like to better understand this. As we point out in the [docs](https://distro.eks.amazonaws.com/#releases) the details of the release are represented [using CRDs](https://distro.eks.amazonaws.com/crds/releases.distro.eks.amazonaws.com-v1alpha1.yaml), for example for the launch release it's https://distro.eks.amazonaws.com/kubernetes-1-18/kubernetes-1-18-eks-1.yaml and so forth.

Are you suggesting to extend the CRD?

**Comment by lancehunt** [comment]
Thanks for pointing this out.  I was looking for something different so this was an unexpected approach.

So, I think I understand what you are saying, but can you clarify a few things about this?

Specifically, will EKS clusters (not self-deployed via kops) have this CRD pre-installed or is this always an optional component?

Does this CRD actually do anything (have an associated operator) or is it just informational?

Is there somewhere that documents non-container dependencies like the Node AMI's and the specific Amazon Linux 2 version or kernel version?

One last comment, there are times when I need to be able to share this information with non-kubernetes users such as our AppSec of InfoSec teams for informational or auditing purposes. It would be nice if there was an easy way to visualize and share this information when them in a more friendly manner.  

Any ideas on this? I could probably just transform the yaml into a UI rendering much like a swagger-spec converted to a UI..

**Comment by lancehunt** [comment]
I think my preference would be for that CRD to be visualized within the Docs for all of the currently supported EKS versions.

**Comment by micahhausler** [comment]
Thanks for opening this issue!

> Specifically, will EKS clusters (not self-deployed via kops) have this CRD pre-installed or is this always an optional component?
> Does this CRD actually do anything (have an associated operator) or is it just informational?

The CRDs are not pre-installed anywhere, and are always an optional component. They are just informational and we don't have an operator for those. The CRD generation code and API code  is in our [aws/eks-distro-build-tooling](https://github.com/aws/eks-distro-build-tooling/tree/main/release) repository or [browse the API Docs on pkg.go.dev](https://pkg.go.dev/github.com/aws/eks-distro-build-tooling/release@v0.0.0-20201211225747-07f05a470de8/api/v1alpha1#Release)

Is an operator something you'd use? What is your deployment target and use case?

> Is there somewhere that documents non-container dependencies like the Node AMI's and the specific Amazon Linux 2 version or kernel version?

The EKS Distro of Kubernetes doesn't include any AMIs, as the focus is the versions of Kubernetes and associated ecosystem dependencies. If you're interested in the AL2/container runtime/kernel version included in the [EKS Optimized AMI](https://github.com/awslabs/amazon-eks-ami), I don't believe that is something they document today, but I'd suggest opening an issue on that repo if that is something you're interested in

> One last comment, there are times when I need to be able to share this information with non-kubernetes users such as our AppSec of InfoSec teams for informational or auditing purposes. It would be nice if there was an easy way to visualize and share this information with them in a more friendly manner.
> Any ideas on this? I could probably just transform the yaml into a UI rendering much like a swagger-spec converted to a UI..

Are you more interested in a browsable schema (such as the swagger spec) or the release content? Would auto-generating markdown files for our docs containing a codeblock with the release CRD manifest be helpful as a first-step?

**Comment by lancehunt** [comment]
> Is an operator something you'd use? What is your deployment target and use case?
I'm not interested in an operator, just confirming that there is no functional aspect to this CRD.

I'm mostly interested in this topic and repo because I use EKS and have wished for more transparency about what it actually is.  This repo and docs and this CRD is sufficient to cover 90% of my needs.  I do like the CRD approach and would love to see that included in EKS as a queryable resource for EKS admins without having to self-deploy the Releases when we update our customers. 

> The EKS Distro of Kubernetes doesn't include any AMIs, as the focus is the versions of Kubernetes and associated ecosystem dependencies.
I was aware of the https://github.com/awslabs/amazon-eks-ami repo but I dont see the two being as unrelated.  For a given EKS/kubernetes version there is a corresponding AMI and without that AMI the EKS-Distro isnt useful - unless someone wants to create their own AMI's perhaps.

I agree they should document the AL2/container runtime/kernel version in that repo, but if you are documenting EKS Distro then there needs to be some way to connect the dots to all dependencies.

> Would auto-generating markdown files for our docs containing a codeblock with the release CRD manifest be helpful as a first-step?
It might be a good start, but thats a long yaml file so not sure how valuable it would be.   I think a pivoting the yaml into a tabular view simply showing the asset, image, and version would be easier to consume.

**Comment by lancehunt** [comment]
Thanks for listening to all of this! 

Its hard for me to judge what everyone would want vs what I want (and my use-case isnt the world's), so I understand your need to figure out that balance as well as to create something that you can sustain.

**Comment by micahhausler** [comment]
> I was aware of the https://github.com/awslabs/amazon-eks-ami repo but I dont see the two being as unrelated. For a given EKS/kubernetes version there is a corresponding AMI and without that AMI the EKS-Distro isnt useful - unless someone wants to create their own AMI's perhaps.

> I agree they should document the AL2/container runtime/kernel version in that repo, but if you are documenting EKS Distro then there needs to be some way to connect the dots to all dependencies.

The ask makes sense, the challenge here is that the EKS Optimized AMI and EKS Anywhere (launching 2021) might have different Linux OS's, kernel versions, and container runtimes, and so tying the Distro to just one of each of those dimensions is de-scoped for now. The goal of the Distro is to be an upstream for both the EKS service and EKS Anywhere.

> I think a pivoting the yaml into a tabular view simply showing the asset, image, and version would be easier to consume.

I think that would look great. I'm not a frontend guy, and pull requests are welcome here :)

**Comment by TerryHowe** [comment]
Sorry, @lancehunt I should have given you more time to review PR #85 If we are far off the mark there, reopen or create a new issue that highlights the missing pieces. We'll enhance this as we go, but I wanted to get something out there. Thanks

**Comment by lancehunt** [comment]
I like what you did here. The new explanation adds enough clarity to address the reasons I opened this issue.  Mission accomplished.

**Comment by lancehunt** [comment]
@micahhausler One last note about this:
>The goal of the Distro is to be an upstream for both the EKS service and EKS Anywhere.

I didnt get that by reading the EKS-D overview (EKS Distro home page)

This is a crucial perspective on this repo so that intro could be improved to say something as concise.  Maybe literally what you said above.
